format/fit/fit.go

```diff
@@ -0,0 +1,343 @@
+		}
+	})
+	if drc.hasDeveloperFields {
+		numDevFields := d.FieldU8("devFields")
+
+		d.FieldArray("devFieldDefinitions", func(d *decode.D) {
+			for i := numFields; i < (numDevFields + numFields); i++ {
+				d.FieldStruct("devFieldDefinition", func(d *decode.D) {
+					fieldDefNo := d.FieldU8("fieldDefNo")
+					size := d.FieldU8("size")
+					devDataIdx := d.FieldU8("devDataIdx")
+
+					fDefLookup, isSet := dmfd[devDataIdx][fieldDefNo]
+
+					if isSet {
+						var foundName = fDefLookup.Name
+						lmfd[drc.localMessageType][i] = mappers.FieldDef{Name: foundName, Type: fDefLookup.Type, Size: size, Unit: fDefLookup.Unit, Scale: fDefLookup.Scale, Offset: fDefLookup.Offset}
+					} else {
+						var foundName = fmt.Sprintf("UNKNOWN_%d", fieldDefNo)
+						lmfd[drc.localMessageType][i] = mappers.FieldDef{Name: foundName, Type: "UNKNOWN", Size: size, Format: "unknown"}
+					}
+				})
+			}
+		})
+	}
+
+}
+
+func ensureDevFieldMap(dmfd devFieldDefMap, devIdx uint64) {
+	_, devIsSet := dmfd[devIdx]
+
+	if !devIsSet {
+		dmfd[devIdx] = make(map[uint64]mappers.FieldDef)
+	}
+}
+
+func fieldUint(fieldFn func(string, ...scalar.UintMapper) uint64, expectedSize uint64, fDef mappers.FieldDef, uintFormatter scalar.UintFn, fdc *fileDescriptionContext) {
+	var val uint64
+
+	if fDef.Size != expectedSize {
+		arrayCount := fDef.Size / expectedSize
+		for i := uint64(0); i < arrayCount; i++ {
+			fieldFn(fmt.Sprintf("%s_%d", fDef.Name, i), uintFormatter)
+		}
+	} else {
+		if uintFormatter != nil {
+			val = fieldFn(fDef.Name, uintFormatter)
+		} else {
+			val = fieldFn(fDef.Name)
+		}
+
+		// Save developer field definitions
+		switch fDef.Name {
+		case "developer_data_index":
+			fdc.devIdx = val
+		case "field_definition_number":
+			fdc.fDefNo = val
+		case "fit_base_type_id":
+			fdc.typ = mappers.TypeDefMap["fit_base_type"][val].Name
+		case "native_field_num":
+			fdc.nativeFieldNo = val
+		case "native_mesg_num":
+			fdc.nativeMsgNo = val
+		}
+	}
+}
+
+func fieldSint(fieldFn func(string, ...scalar.SintMapper) int64, expectedSize uint64, fDef mappers.FieldDef, sintFormatter scalar.SintFn) {
+	if fDef.Size != expectedSize {
+		arrayCount := fDef.Size / expectedSize
+		for i := uint64(0); i < arrayCount; i++ {
+			fieldFn(fmt.Sprintf("%s_%d", fDef.Name, i), sintFormatter)
+		}
+	} else {
+		if sintFormatter != nil {
+			fieldFn(fDef.Name, sintFormatter)
+		} else {
+			fieldFn(fDef.Name)
+		}
+	}
+}
+
+func fieldFloat(fieldFn func(string, ...scalar.FltMapper) float64, expectedSize uint64, fDef mappers.FieldDef, floatFormatter scalar.FltFn) {
+	if fDef.Size != expectedSize {
+		arrayCount := fDef.Size / expectedSize
+		for i := uint64(0); i < arrayCount; i++ {
+			fieldFn(fmt.Sprintf("%s_%d", fDef.Name, i), floatFormatter)
+		}
+	} else {
+		fieldFn(fDef.Name)
+	}
+}
+
+func fieldString(d *decode.D, fDef mappers.FieldDef, fdc *fileDescriptionContext) {
+	val := d.FieldUTF8NullFixedLen(fDef.Name, int(fDef.Size), scalar.StrMapSymStr{"": "[invalid]"})
```

Comment:
I wonder if this should map "" to the description "Invalid" instead as i guess the string "[invalid]" is not useful query-wise?